pkg/adaptor/mongodb/tailer.go

```diff
@@ -0,0 +1,174 @@
+
+	"github.com/compose/transporter/pkg/client"
+	"github.com/compose/transporter/pkg/log"
+	"github.com/compose/transporter/pkg/message"
+	"github.com/compose/transporter/pkg/message/data"
+	"github.com/compose/transporter/pkg/message/ops"
+	mgo "gopkg.in/mgo.v2"
+	"gopkg.in/mgo.v2/bson"
+)
+
+var (
+	_ client.Reader = &Tailer{}
+)
+
+// Tailer implements the behavior defined by client.Tailer for interfacing with the MongoDB oplog.
+type Tailer struct {
+	reader client.Reader
+	db     string
+
+	oplogTimeout time.Duration
+}
+
+func newTailer(db string) client.Reader {
+	return &Tailer{newReader(db), db, 5 * time.Second}
+}
+
+// Tail does the things
+func (t *Tailer) Read(filterFn client.NsFilterFunc) client.MessageChanFunc {
+	return func(s client.Session, done chan struct{}) (chan message.Msg, error) {
+		oplogTime := timeAsMongoTimestamp(time.Now())
+		log.Infof("setting oplog start timestamp: %d", oplogTime)
+
+		readFunc := t.reader.Read(filterFn)
+		msgChan, err := readFunc(s, done)
+		if err != nil {
+			return nil, err
+		}
+		out := make(chan message.Msg)
+		go func() {
+			defer close(out)
+			for msg := range msgChan {
+				out <- msg
+			}
+
+			var (
+				collection = s.(*Session).mgoSession.DB("local").C("oplog.rs")
+				result     oplogDoc // hold the document
+				query      = bson.M{"ts": bson.M{"$gte": oplogTime}}
+				iter       = collection.Find(query).LogReplay().Sort("$natural").Tail(t.oplogTimeout)
+			)
+			defer iter.Close()
+
+			for {
+				log.With("db", t.db).Infof("tailing oplog with query %+v\n", query)
+				select {
+				case <-done:
+					log.With("db", t.db).Infoln("tailing stopping...")
+					return
+				default:
+					for iter.Next(&result) {
+						if result.validOp() {
+							db, c, _ := splitNamespace(result.Ns)
+							if db != t.db {
+								continue
+							}
+							if !filterFn(c) {
+								continue
+							}
+
+							var (
+								doc bson.M
+								err error
+								op  ops.Op
+							)
+							switch result.Op {
+							case "i":
+								op = ops.Insert
+								doc = result.O
+							case "d":
+								op = ops.Delete
+								doc = result.O
+							case "u":
+								op = ops.Update
+								doc, err = t.getOriginalDoc(result.O2, c, s.(*Session).mgoSession)
+								if err != nil {
+									// errors aren't fatal here, but we need to send it down the pipe
+									log.With("ns", result.Ns).Errorf("unable to getOriginalDoc, %s\n", err)
+									// m.pipe.Err <- adaptor.NewError(adaptor.ERROR, m.path, fmt.Sprintf("tail MongoDB error (%s)", err.Error()), nil)
+									continue
+								}
+							}
+
+							msg := From(op, fmt.Sprintf("%s.%s", t.db, c), data.Data(doc)).(*Message)
+							msg.TS = int64(result.Ts) >> 32
+
+							out <- msg
+							oplogTime = result.Ts
+						}
+						result = oplogDoc{}
+					}
+				}
+
+				if iter.Timeout() {
+					continue
+				}
+				if iter.Err() != nil {
+					log.With("path", t.db).Errorf("error tailing oplog, %s\n", iter.Err())
+					break
```

Comment:
This break will break us out of the whole tailing process, right? We should probably do some retrying like we do for catting.